cpp/src/arrow/compute/kernels/vector_rolling.cc

```diff
@@ -0,0 +1,785 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <functional>
+#include <set>
+#include <type_traits>
+#include <utility>
+#include "arrow/array/array_base.h"
+#include "arrow/array/builder_primitive.h"
+#include "arrow/compute/api_scalar.h"
+#include "arrow/compute/api_vector.h"
+#include "arrow/compute/cast.h"
+#include "arrow/compute/kernels/base_arithmetic_internal.h"
+#include "arrow/compute/kernels/codegen_internal.h"
+#include "arrow/compute/kernels/common_internal.h"
+#include "arrow/compute/registry.h"
+#include "arrow/result.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/bit_block_counter.h"
+#include "arrow/util/bit_util.h"
+#include "arrow/util/logging.h"
+#include "arrow/visit_type_inline.h"
+
+namespace arrow::compute::internal {
+namespace {
+struct RollingOptionsWrapper : public OptionsWrapper<RollingOptions> {
+  explicit RollingOptionsWrapper(RollingOptions options)
+      : OptionsWrapper(std::move(options)) {}
+
+  static Result<std::unique_ptr<KernelState>> Init(KernelContext* ctx,
+                                                   const KernelInitArgs& args) {
+    auto options = checked_cast<const RollingOptions*>(args.options);
+    if (!options) {
+      return Status::Invalid(
+          "Attempted to initialize KernelState from null FunctionOptions");
+    }
+
+    if (options->min_periods > options->window_length) {
+      return Status::Invalid("min_periods must be less than or equal to window_length");
+    }
+
+    if (options->window_length <= 0) {
+      return Status::Invalid("window_length must be greater than 0");
+    }
+
+    if (options->min_periods <= 0) {
+      return Status::Invalid("min_periods must be greater than 0");
+    }
+
+    return std::make_unique<RollingOptionsWrapper>(*options);
+  }
+};
+
+template <typename ArgType>
+struct SumWindow {
+  using OutType = ArgType;
```

Comment:
If you can the parameter `ArrowType` or `DataType` you can use it in place of `ArgType` and `OutType` removing the need for the aliasing and different names for a type that is the same.

Reply:
But if you think some `Window`s might need to differentiate here, keep it as it is.

Reply:
Yeah I would prefer to keep this because Mean has a different out type than arg type

Reply:
Makes sense!